server/src/main/java/com/cloud/network/NetworkServiceImpl.java

```diff
@@ -732,11 +732,10 @@ public NicSecondaryIp allocateSecondaryGuestIP(final long nicId, IpAddresses req
             }
 
             try {
-                if (ipv4Address != null) {
-                    ipaddr = _ipAddrMgr.allocatePublicIpForGuestNic(network, podId, ipOwner, ipv4Address);
-                }
                 if (ipv6Address != null) {
                     ip6addr = ipv6AddrMgr.allocatePublicIp6ForGuestNic(network, podId, ipOwner, ipv6Address);
+                } else {
```

Comment:
I'm a bit confused, the null check for ipv4address is gone and it is not assigned (even when not null) when it isn't but ipv6Address is not null. This means that if both are available ipv6 is the *only one* used. Is that what you intend @ustcweizhou ?
I see how this fixes backwards compatibility btw, just wondering if we would want to allow dualstack!?

Reply:
@DaanHoogland 
the method is called only when add a nic to vm.
https://github.com/apache/cloudstack/blob/master/api/src/main/java/org/apache/cloudstack/api/command/user/vm/AddIpToVmNicCmd.java#L159-L162

at most one ip (ipv4 or ipv6) is passed.

the idea of this PR is, if ipv6 is passed, check and assign an ipv6 address. It ipv4 is passed or no ip is passed, then assign an ipv4 address.

Reply:
👍